examples/shader/automatic_instancing.rs

```diff
@@ -1,51 +1,102 @@
 //! Shows that multiple instances of a cube are automatically instanced in one draw call
 //! Try running this example in a graphics profiler and all the cubes should be only a single draw call.
+//! Also demonstrates how to use `MeshTag` to use external data in a custom material.
 
-use bevy::prelude::*;
+use bevy::{
+    prelude::*,
+    reflect::TypePath,
+    render::{
+        mesh::MeshTag,
+        render_resource::{AsBindGroup, ShaderRef},
+    },
+};
+
+const SHADER_ASSET_PATH: &str = "shaders/automatic_instancing.wgsl";
 
 fn main() {
     App::new()
-        .add_plugins(DefaultPlugins)
+        .add_plugins((DefaultPlugins, MaterialPlugin::<CustomMaterial>::default()))
         .add_systems(Startup, setup)
+        .add_systems(Update, update)
         .run();
 }
 
-/// set up a simple 3D scene
+/// Sets up an instanced grid of cubes, where each cube is colored based on an image that is
+/// sampled in the vertex shader. The cubes are then animated in a spiral pattern.
+///
+/// This example demonstrates one use of automatic instancing and how to use `MeshTag` to use
+/// external data in a custom material. For example, here we use the "index" of each cube to
+/// determine the texel coordinate to sample from the image in the shader.
 fn setup(
     mut commands: Commands,
+    assets: Res<AssetServer>,
     mut meshes: ResMut<Assets<Mesh>>,
-    mut materials: ResMut<Assets<StandardMaterial>>,
+    mut materials: ResMut<Assets<CustomMaterial>>,
 ) {
-    // camera
+    // We will use this image as our external data for our material to sample from in the vertex shader
+    let image = assets.load("branding/icon.png");
+
+    // Our single mesh handle that will be instanced
+    let mesh_handle = meshes.add(Cuboid::from_size(Vec3::splat(0.01)));
+
+    // Create the custom material with a reference to our texture
```

Comment:
```suggestion
    // Create the custom material with a reference to our texture
    // Automatic instancing works with any Material, including the `StandardMaterial`.
    // This custom material is used to demonstrate the optional `MeshTag` feature.
```

Maybe add something like this too? Just to make sure people understand there's nothing special about the custom material in terms of automatic instancing.